File: library/src/main/java/org/incendo/interfaces/properties/InterfaceProperty.java
package org.incendo.interfaces.properties;

public final class InterfaceProperty<T> extends DelegateTrigger {

    public static <T> InterfaceProperty<T> interfaceProperty(final T defaultValue) {
        return new InterfaceProperty<>(defaultValue);
    }

    private T value;

    public InterfaceProperty(final T defaultValue) {
        this.value = defaultValue;
    }

    public T value() {
        return this.value;
    }

    public void value(final T newValue) {
        T oldValue = this.value;
        this.value = newValue;

        if (oldValue != newValue) {
            this.trigger();
        }
    }
}

File: next/src/main/kotlin/org/incendo/interfaces/next/properties/InterfaceProperty.kt
package org.incendo.interfaces.next.properties

import kotlin.properties.ObservableProperty
import kotlin.reflect.KProperty

public class InterfaceProperty<T>(
    defaultValue: T
) : ObservableProperty<T>(defaultValue), Trigger by DelegateTrigger() {

    override fun afterChange(property: KProperty<*>, oldValue: T, newValue: T) {
        if (oldValue != newValue) {
            trigger()
        }
    }
}
